### ISSUE TITLE
Wayland Capturing

### ISSUE THREAD

**Issue by routerino** [Ylianst/MeshCentral#2719]
So the current status is that wayland is a nonstarter in meshcentral due to its restrictions on traditianal frame capture.

I thought it would be worth putting out that [OBS](https://www.reddit.com/r/linux/comments/nq202v/obs_studio_27_released_with_native_wayland_and/) has added support for wayland, and their efforts may be applicable to meshcentral.

If there's an option to reuse code for real time capture, then great. I don't think wayland is disappearing any time soon.

**Comment by Ylianst** [comment]
Good observation. Bryan needs to look into this.

**Comment by bt90** [comment]
Teamviewer does also provide experimental Wayland support now implemented via xdg-desktop-portal:

https://community.teamviewer.com/English/discussion/122410/

**Comment by veitw** [comment]
The wayvnc VNC server for Wayland also exists and should have a much less complex code base than OBS to figure out the workings:

https://github.com/any1/wayvnc

**Comment by Mannshoch** [comment]
Gnome support RDP on Wayland out of the box. https://wiki.gnome.org/Projects/Mutter/RemoteDesktop

**Comment by bt90** [comment]
Would it be possible to use xdg-desktop-portal? This is what Teamviewer and Rustdesk are using:

https://github.com/rustdesk/rustdesk/pull/932

**Comment by DaanSelen** [comment]
What is the status on Wayland support?

**Comment by si458** [comment]
closing as duplicate of #115, so please follow there for update

**Comment by hobbes** [comment]
the right duplicate issue is https://github.com/Ylianst/MeshAgent/issues/115 :-)

**Comment by si458** [comment]
sorry reopening, so it reminds users about the issue and they can link to the meshagent repo for code etc

**Comment by si458** [comment]
workaround for users for future, will update docs - https://github.com/Ylianst/MeshCentral/issues/2998#issuecomment-894554156

1. Open /etc/gdm/custom.conf or /etc/gdm3/custom.conf
2. Uncomment: WaylandEnable=false.
3. Add the following line to the [daemon] section: DefaultSession=gnome-xorg.desktop
4. restart machine

**Comment by marclaporte** [comment]
@lsowen, @timwhite, @nosklo, @flifloo, @artemklevtsov, @laramy2020, @fcayre, @zroug, @LambArchie, @Mannshoch, @Attacktive, @bt90, @weitw, @DaanSelen, @hobbes

We'd love to add this but it's a lot of work. So we are looking for a developer to contribute or a financial sponsor. Any ideas?

You can reach out in private to me and/or @si458 to discuss.

Thanks!

**Comment by Saik0Shinigami** [comment]
@marclaporte @si458 what kind of financial sponsorship are we talking here? My hands are quite full on my own projects (and I was never a C++ dev). However, I could throw some (small company) money into it. I love Mesh... I hate that I can't use it on half my systems now for GUI access. This issue has been outstanding long enough that I'd like to see it squashed.

**Comment by marclaporte** [comment]
@Saik0Shinigami We don't know the precise amount because we can't ask someone to evaluate until/unless we have a realistic way to fund it. But for sure, it is a lot of money for a small company, so ideally we have a bunch of people and organizations that co-sponsor.

Also, we can apply for a grant from organizations like NLnet:
https://nlnet.nl/news/2024/20240601-call.html

We need a volunteer to craft the grant request and coordinate the project.

**Comment by hobbes** [comment]
> @Saik0Shinigami We don't know the precise amount because we can't ask someone to evaluate until/unless we have a realistic way to fund it. 

I think it's the other way round: there are many ways to fund that kind of development (you cited two, there are probably more), but there's no way to chose, devise a strategy, or draft a grant request if we don't at least have a rough idea of how much it represents.

Now if I understand the problem correctly, there is no way with wayland to just capture the screen and/or control the inputs like the agent does with X. So adding wayland support would mean significant modifications to the architecture of the agent, like adding new behaviours (for example, asking explicit consent from the user, interacting with portals,...) in addition to the pure wayland support. 

That's what needs investigating first, if possible by someone familiar with the wayland ecosystem: how would a remote support solution present itself to the users of both sides of the solution (support staff on one side, and machine users on the other side) with the wayland stack. What is possible (is unattended remote control even possible?), how, and with what tools.

Then, with that information, someone familiar with the MC stack can evaluate what needs to be done agentside to implement it.

Both of these can be financed via sponsoship, but we need an estimate...

**Comment by marclaporte** [comment]
> there's no way to chose, devise a strategy, or draft a grant request if we don't at least have a rough idea of how much it represents.

You make a good point. Perhaps someone could even write up a grant request for the evaluation project itself. Like a technical project appraisal: https://en.wikipedia.org/wiki/Project_appraisal

So we need a volunteer to take the lead here.

**Comment by hobbes** [comment]
sorry for the delay, I fell ill in the meantime :-)

ok I don't think that it's possible do draft a grand request or anything since we don't know who could do it, how long it would take, or anything. We first need to beat it up a bit. 

The first thing to know is what is possible with wayland. 

capturing is possible: firefox does it for desktop sharing in remote meeting solutions like jitsi, gnome does it for screen recording, obs studio does it,... but for all these, the user needs to allow the app to do it. 

I think that all these use the xdg-desktop-portal infrastructure. After reading a bit in the doc, probably what would be used by meshcentral is https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.RemoteDesktop.html

Now I think that these require the user to give permission, so that wouldn't be possible for the current agent behaviour, but ok for something like the meshcentral assistant.

For agent behaviour, now is the time to hunt for wayland people :-)

If anyone here knows someone who might be familiar with the wayland ecosystem, can you ask them to come here, or ask theme those questions: 

- is it possible with wayland to capture the whole screen as soon as the display manager (gdm, lightdm,...) is there
- if not, are there existing remote support/management solutions working with wayland that we could analyse ?

I don't know anyone familiar with wayland, but I'll ask on the fediverse. Don't hesitate to ring your social networks :-)

**Comment by nekohayo** [comment]
> * is it possible with wayland to capture the whole screen as soon as the display manager (gdm, lightdm,...) is there
> * if not, are there existing remote support/management solutions working with wayland that we could analyse ?

It will be / is becoming possible with GNOME 46.x-47.x. and newer.

From an architectural standpoint, GNOME is finally doing remote desktop "properly" (i.e. fairly safely and efficiently) now, thanks to Wayland and the huge amount of work that the GNOME Remote Desktop (and coincidentally GNOME Shell, Mutter, GDM, GStreamer, etc.) developers have been doing. It's all based on RDP for performance (you can't say "VNC" and "fast" in the same sentence and look at me with a straight face :)

You can get see the big picture by looking at this collection of interlocking implementation parts, with their related tickets and merge requests:

* Basic RDP-based remote desktop for "already running and unlocked" sessions (i.e. the "tech support" scenario): already implemented a long time ago, somewhere in 2022 I think. Don't remember which ticket it was.
* [Graphical remote login through GDM](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/92) (gnome-remote-desktop meta issue 92)
* [Support headless sessions](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/90) (gnome-remote-desktop meta issue 90)
* [Support hybrid headless/headfull sessions](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/91) (gnome-remote-desktop meta issue 91)

See also the Red Hat Desktop Team lead's blog posts for additional historical context (`Ctrl+F` for "remote" in those blog posts):

* 2022: "[Update from the world of Fedora Workstation](https://blogs.gnome.org/uraeus/2022/12/14/update-from-the-world-of-fedora-workstation/)"
* 2023: "[Fedora Workstation 39 and beyond](https://blogs.gnome.org/uraeus/2023/11/29/fedora-workstation-39-and-beyond/)"
* 2024: "[Fedora Workstation 40 – what are we working on](https://blogs.gnome.org/uraeus/2024/03/28/fedora-workstation-40-what-are-we-working-on/)"

**Comment by hobbes** [comment]
> > * is it possible with wayland to capture the whole screen as soon as the display manager (gdm, lightdm,...) is there
> > * if not, are there existing remote support/management solutions working with wayland that we could analyse ?
> 
> It will be / is becoming possible with GNOME 46.x-47.x. and newer.

wow, thanks a lot for that detailed answer :-)

...

> You can get see the big picture by looking at this collection of interlocking implementation parts, with their related tickets and merge requests:
> 
>     * Basic RDP-based remote desktop for "already running and unlocked" sessions (i.e. the "tech support" scenario): already implemented a long time ago, somewhere in 2022 I think. Don't remember which ticket it was.

Ok, I see that now in my control center, I can activate remote desktop for my running unlocked session

>     * [Graphical remote login through GDM](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/92) (gnome-remote-desktop meta issue 92)
> 
>     * [Support headless sessions](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/90) (gnome-remote-desktop meta issue 90)
> 
>     * [Support hybrid headless/headfull sessions](https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/91) (gnome-remote-desktop meta issue 91)

and these too, I can now enable remote graphical login. However, if I understand correctly, this only enables the creation of a remote session which would be invisible to a user on the physical machine. A bit like it works on windows for rdp. 

That's already a big part of what is needed, maybe it's enough to start ?

Actually, if that's rdp based, it might already work with current agents... but my mc server is not in the same network as my machines, so rdp does not work for me, so I can't test.

@marclaporte , @si458 : would it be conceivable to implement something like rdp (or just tcp port) forwarding between the agent and the mc server when the agent is not directly reachable by the server ? Because if yes, problem largely solved for gnome and any desktop that implements remote desktop :-) 

And bonus: that would give rdp possibility also on other platforms when the managed machines are not directly reachable by the mc server (like behind nat).

**Comment by Fuseteam** [comment]
there is now also a screen capture wayland protocol fwiw https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124

**Comment by marclaporte** [comment]
Also discussed here: https://www.reddit.com/r/MeshCentral/comments/1bgjyna/wayland_blues_what_are_people_doing_now/

**Comment by marclaporte** [comment]
I went to XDC 2024 in October 2024: https://indico.freedesktop.org/event/6/

There were several Wayland experts so I asked for advice. The suggestion was to look into:
- vncforwlroots as a reference
- Pipewire multimedia framework
- ei (emulated input)

They explained it is not a difficult task because it has been done before. So I take it, not difficult for a Wayland expert :-)

**Comment by tomsik-radek** [comment]
I hope one day Wayland support comes (not a dev in the slightest, just a user). RHEL10 (and with EnterpriseLinux 10) is going all in on Wayland, as far as totally removing xorg

**Comment by DaanSelen** [comment]
We need to discuss funding in the next meetings.

**Comment by marclaporte** [comment]
> would it be conceivable to implement something like rdp (or just tcp port) forwarding between the agent and the mc server when the agent is not directly reachable by the server

I don't know. We need a dev to volunteer, however it's done...

**Comment by hobbes** [comment]
please do not close old valid issues just because nobody from the current team knows how to solve them. The issue are still valid, and someone with the right know-how could stumble on it...

**Comment by DaanSelen** [comment]
> please do not close old valid issues just because nobody from the current team knows how to solve them. The issue are still valid, and someone with the right know-how could stumble on it...

Yeah we are not closing them, we are now getting used to marking them accordingly. As many issues are no longer relevant.

**Comment by acrey1** [comment]
This has acquired a new level of importance / urgency, as Canonical has now dropped support for GNOME on Xorg in Ubuntu 25.10.

https://discourse.ubuntu.com/t/ubuntu-25-10-drops-support-for-gnome-on-xorg/62538

**Comment by DaanSelen** [comment]
> This has acquired a new level of importance / urgency, as Canonical has now dropped support for GNOME on Xorg in Ubuntu 25.10.
> 
> https://discourse.ubuntu.com/t/ubuntu-25-10-drops-support-for-gnome-on-xorg/62538

We are waiting on volunteers or people doing it for payment I think.

**Comment by hobbes** [comment]
> This has acquired a new level of importance / urgency, as Canonical has now dropped support for GNOME on Xorg in Ubuntu 25.10.
> 
> https://discourse.ubuntu.com/t/ubuntu-25-10-drops-support-for-gnome-on-xorg/62538

note that it's now possible to enable RDP desktop sharing in gnome, which should alleviate at least part of the need for native wayland capturing, *but* it's broken with meshcentral at the moment #6384 :-)

Diagnosing #6384 has been on my tasklist for some times now, but I have more pressing matters to attend to, so don't hold your breath...

**Comment by Fuseteam** [comment]
> This has acquired a new level of importance / urgency, as Canonical has now dropped support for GNOME on Xorg in Ubuntu 25.10.
> 
> https://discourse.ubuntu.com/t/ubuntu-25-10-drops-support-for-gnome-on-xorg/62538

I kinda wonder if we can contact canonical about ways to do this

**Comment by acrey1** [comment]
https://www.phoronix.com/news/Fedora-43-Wayland-Only-GNOME

Fedora also.  That's (arguably) the two biggest distros...

**Comment by jontylovell** [comment]
It seems that wayvnc is now available in most linux distros. Can we hook into wayvnc to remote control modern linux desktops. The Raspberry Pi Foundation do this in Raspberry Pi OS. There is a standard feature called Raspberry Pi Connect which is very like Mesh Central. It allows terminal or full screen sharing even behind firewalls. All of the screen sharing runs via wayvnc.

**Comment by si458** [comment]
@jontylovell this is already possible.
enable vnc, set ya user/pass etc (check u can connect with a normal vnc viewer)
install the meshagent, click the web-vnc option, connect, user/pass, away u go

**Comment by jontylovell** [comment]
@si458 This has always failed for me. I can directly VNC to these devices, including the mesh central server, on my LAN. When I try the web-VNC to any of these devices it always returns almost immediately with "Failed to connect to server". Under the advanced settings everything looks correct. Is there something I am missing?

**Comment by si458** [comment]
@jontylovell no its always worked for me and others?
what OS/version are u using?
also what VNC app/version too?

**Comment by jontylovell** [comment]
@si458 I am running Mesh Central on Debian Bookworm. Most of my devices are running Bookworm. I have a couple of Windows 11 devices that I am accessing the console from. I am running the latest wayvnc on all of the Debian devices and I can VNC onto them directly without any problems, including from the Mesh Central server (which is not a docker image by the way, I have installed locally). The RDP connect to the Windows devices works and the regular connect works for all devices (although it says it can't connect to the wayland back end and to switch to xorg). But the web-RDP and Web-VNC both fail.

**Comment by si458** [comment]
@jontylovell ok just wanted to clarify for my own mind (brain cabbaged really need to see a doctor)

so raspberry pi, debian bookworm, all latest updates etc, wayvnc as the vnc application
so this means you are using wayland and NOT X
all correct?

have you tried using the build in realvnc provided by raspberry pi?
as i know this works as i use it

as for the RDP, yes im already aware RDP doesnt work with linux, doesnt matter is wayland or X,
needs looking into as its a linux thing (an issue already open for this)

**Comment by dustojnikhummer** [comment]
> It seems that wayvnc is now available in most linux distros. Can we hook into wayvnc to remote control modern linux desktops. The Raspberry Pi Foundation do this in Raspberry Pi OS. There is a standard feature called Raspberry Pi Connect which is very like Mesh Central. It allows terminal or full screen sharing even behind firewalls. All of the screen sharing runs via wayvnc.

I thought WayVNC was limited to wlroots, and therefore will not work with regular Wayland+Gnome/KDE/Cinnamon?

**Comment by dustojnikhummer** [comment]
> [@jontylovell](https://github.com/jontylovell) ok just wanted to clarify for my own mind (brain cabbaged really need to see a doctor)
> 
> so raspberry pi, debian bookworm, all latest updates etc, wayvnc as the vnc application so this means you are using wayland and NOT X all correct?
> 
> have you tried using the build in realvnc provided by raspberry pi? as i know this works as i use it
> 
> as for the RDP, yes im already aware RDP doesnt work with linux, doesnt matter is wayland or X, needs looking into as its a linux thing (an issue already open for this)

While I do think that current FreeRDP (on Linux side, not Mesh) is underbaked (and won't be usable for a decade because RHEL10), temporary workaround could be allowing MeshCentral Router mapping to 3389 and be recognized by MSTSC, as was suggested in #6907, because that mostly works

![Image](https://github.com/user-attachments/assets/6e9a26ed-09dd-481b-b6d3-c83c45c8bd13)

**Comment by Fuseteam** [comment]
I have found some documentation: https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.InputCapture.html
I believe this portal allow capturing input devices on wayland
there also seems to be a portal specifically for remote desktop:
https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.RemoteDesktop.html

**Comment by si458** [comment]
we are making progress with this recently!
im hoping if more people can test https://github.com/Ylianst/MeshAgent/pull/351 we could get it merged!

**Comment by hobbes** [comment]
> im hoping if more people can test [Ylianst/MeshAgent#351](https://github.com/Ylianst/MeshAgent/pull/351) we could get it merged!

sure, but I'd need instructions how to create (or easier, download) an agent binary with that pr, and how I deploy it on some ok my linux hosts... I'm not a developer, so bear with me :-)

**Comment by katamadone** [comment]
I'd be also in for a test.

**Comment by Rambomst** [comment]
@hobbes @katamadone the latest work can be found at https://github.com/Ylianst/MeshAgent/pull/351.

After checking out that branch:

1. Build it: `make clean && make linux ARCHID=6`
2. Run it in the foreground with debug logging: `MESH_DISABLE_UPDATE=1 MESH_KVM_DRM_DEBUG=2 ./meshagent_x86-64 --disableUpdate=1 2>&1 | tee /tmp/meshkvmdebug.log`
3. Connect.

If anythings wrong, attach /tmp/meshkvmdebug.log

**Comment by dustojnikhummer** [comment]
> @hobbes @katamadone the latest work can be found at https://github.com/Ylianst/MeshAgent/pull/351.
> 
> After checking out that branch:
> 
> 1. Build it: `make clean && make linux ARCHID=6`
> 2. Run it in the foreground with debug logging: `MESH_DISABLE_UPDATE=1 MESH_KVM_DRM_DEBUG=2 ./meshagent_x86-64 --disableUpdate=1 2>&1 | tee /tmp/meshkvmdebug.log`
> 3. Connect.
> 
> If anythings wrong, attach /tmp/meshkvmdebug.log

I'm not a dev so forgive the question, but is this just building the agent or also the server?

**Comment by Rambomst** [comment]
@dustojnikhummer just the agent, no server changes required.

**Comment by Fuseteam** [comment]
do we need a specific version of the server?

**Comment by Rambomst** [comment]
> do we need a specific version of the server?

@Fuseteam probably not

**Comment by raffaelebacchi-netizen** [comment]
Hi all, thanks to AI (antigraviti cli) i made a C visual studio code project capable of streaming de wayland desktop (mjpeg). Built and tested on a VM Ubuntu desktop 26. you can run and view desktop on browser. See readme for details. Hope it can help.
https://drive.google.com/drive/folders/10yNxVz3-qWqlSfvBv1EZ2T1lLd6Lq1lU?usp=sharing

**Comment by Fuseteam** [comment]
@raffaelebacchi-netizen if you scroll up you'll see what it might already be possible— one simply need to test it

**Comment by raffaelebacchi-netizen** [comment]
@Fuseteam Sorry di not see the progress recently done. It looks like it is almost ready to be released

**Comment by kasperlaessoe** [comment]
I have been using this and tested it. Works great on omarchy :) Will test on Fedora.
